zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+// Copyright 2023 Forecasting Technologies LTD.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+#![doc = include_str!("../README.md")]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+mod consts;
+mod macros;
+mod math;
+pub mod mock;
+mod tests;
+mod traits;
+mod types;
+mod weights;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+mod pallet {
+    use crate::{
+        consts::MAX_ASSETS,
+        math::{Math, MathOps},
+        traits::{pool_operations::PoolOperations, DistributeFees, LiquiditySharesManager},
+        types::{FeeDistribution, Pool, SoloLp},
+        weights::*,
+    };
+    use alloc::collections::BTreeMap;
+    use core::marker::PhantomData;
+    use frame_support::{
+        dispatch::DispatchResultWithPostInfo,
+        ensure,
+        pallet_prelude::StorageMap,
+        require_transactional,
+        traits::{Get, IsType, StorageVersion},
+        transactional, PalletId, Twox64Concat,
+    };
+    use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+    use orml_traits::MultiCurrency;
+    use sp_runtime::{
+        traits::{AccountIdConversion, CheckedAdd, CheckedSub, Saturating, Zero},
+        DispatchError, DispatchResult, SaturatedConversion,
+    };
+    use zeitgeist_primitives::{
+        constants::{BASE, CENT},
+        math::fixed::{bdiv, bmul},
+        traits::{CompleteSetOperationsApi, DeployPoolApi, ZeitgeistAssetManager},
+        types::{Asset, MarketStatus, MarketType, ScalarPosition, ScoringRule},
+    };
+    use zrml_market_commons::MarketCommonsPalletApi;
+
+    // These should not be config parameters to avoid misconfigurations.
+    pub(crate) const MIN_SWAP_FEE: u128 = BASE / 1_000; // 0.1%.
+    pub(crate) const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+    pub(crate) const MAX_SPOT_PRICE: u128 = BASE - CENT / 2;
+    pub(crate) const MIN_SPOT_PRICE: u128 = CENT / 2;
+    pub(crate) const MIN_LIQUIDITY: u128 = BASE;
```

Comment:
That's a good idea. However, I would ask that we make this a task for the full AMM 2.0 implementation. Right now, there's still some AMM logic found in the `do_*` functions and it would require some restructuring, plus I'm not 100% sure how exactly the solution for multimarkets looks, so...

Reply:
https://github.com/zeitgeistpm/zeitgeist/issues/1106